Adds split method to port class

mvillius opened the pull request:
Pretty self-explanatory

Commit:
Adds split method to port class

rlescanne commented on HFSSdrawpy/core/port.py:
```diff
@@ -183,3 +184,49 @@ def rotate_ports(ports, angle):
+            raise ValueError('Split ports must belong to '+str(self.subnames))
+
+        self.save = {'name':self.name, 'pos':self.pos, 'widths':self.widths,
+                     'subnames':self.subnames, 'offsets':self.offsets}
```
This is not needed here (I think). The save is only required when performing a brief modification of the port (such as when drawing adaptors, the port is temporarily shifted and resized to the other port specs).

rlescanne commented on HFSSdrawpy/core/port.py:
```diff
@@ -183,3 +184,49 @@ def rotate_ports(ports, angle):
             posy = port.pos[0]*np.sin(rad)+port.pos[1]*np.cos(rad)
             port.pos = Vector([posx, posy])
+
+    def split(self, subnames=[], gap=None):
```
Je ne sais plus pourquoi mais on n'a pas le droit d'initialiser par défaut à une liste en python. Il vaut mieux mettre un None et propager cette modification dans ton code.

rlescanne commented on HFSSdrawpy/core/port.py:
```diff
@@ -183,3 +184,49 @@ def rotate_ports(ports, angle):
+        if not subnames:
+            subnames = self.subnames
+        elif subnames == -1:
+            subnames = self.subnames[:-1]
```
I guess this assumes the last subport defined is the outline/gap. This assumption is not ideal but I guess this is the best solution for now.

rlescanne commented on HFSSdrawpy/core/port.py:
```diff
@@ -183,3 +184,49 @@ def rotate_ports(ports, angle):
+            name     = self.name+'_'+sub
+            pos      = self.pos + Vector([0,self.offsets[isub]]).rot(self.ori)
+            widths   = [self.widths[isub], self.widths[isub]+2*gap]
+            layers   = [TRACK, GAP]
```
To add a little bit of flexibility here, I would replace `TRACK` by the layer of the corresponding subport. I agree that we cannot give this flexibility to `GAP`.

rlescanne reviewed (changes requested):
Qqs petites remarques

Commit:
Pull request #31
Answer to Raph's heavy complains

MVillius commented:
Done

Commit:
Updating split method so that one can retrieve several subports from a multiport and recover the updated multiport

Commit:
Increasing tolerance for tube stuff

MVillius commented:
I added the possibility to split a multiport (non-CPW) into several multiports, and since the previous Pull Request had not been closed it has been included in the existing one. What do you think Raph?

Commit:
Improve draw_cable for mask with contraint_port and split multiports

Commit:
- Correct bug so that mask can cover contraint_ports
- Double hack on the draw_cable body method and the split port method so that mask can cover split multiport

pierreguilmin committed:
Merge branch 'master' into marius

pierreguilmin commented on HFSSdrawpy/core/port.py:
```diff
@@ -105,9 +105,11 @@ def compare(self, other, pm, slope=0.5):
             offset1 = self.offsets[ii]
-            offset2 = other.offsets[ii]
+
+            offset2 = -other.offsets[ii]
```
@MVillius do you think I can merge this modification into master or is it just a specific fix for your work?

mvillius replied:
This modification makes your life easier when you want to draw adaptors with off-centered multiports. I think I am the only one to do this so you should be able to merge this modification into master seamlessly

mvillius commented on HFSSdrawpy/interfaces/gds_modeler.py:
```diff
@@ -4,7 +4,7 @@
 from ..utils import Vector, parse_entry, val
 
-TOLERANCE = 1e-8  # for arcs
+TOLERANCE = 1e-9 # for arcs
```
This is the new standard for everybody in the lab, we should keep 1e-9
